### PR TITLE
[Issue-25288]: build new social media tile preview tool for Twitter

### LINKED ISSUE #25288: Build New Social Media Tile Preview Tool for Twitter
### Parent Issue

_No response_

### Task

Build New Social Media Tile Preview Tool for Twitter

### Proposed Objective

Same as Parent Issue

### Proposed Priority

Same as Parent Issue

### Acceptance Criteria

See notes here:

* Social Media Tile & Search Engine Results Page Preview Tool Messages | dotCMS - Google Docs || https://docs.google.com/document/d/1O8PMni7FzD9YmIHdz-De43MnHpO--0_W49ZwYxLt9DM/edit

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

Demo - https://www.figma.com/proto/jyu7b1wgL16iV7yOoNdgu9/SEO?page-id=1%3A65&type=design&node-id=298-6800&viewport=-1440%2C-118%2C0.37&scaling=scale-down&starting-point-node-id=372%3A9190 

Design - https://www.figma.com/file/jyu7b1wgL16iV7yOoNdgu9/SEO?type=design&node-id=298-6961&t=kzAG714FzCpMdHYC-0 

Sanity Check Requirements - https://socialsharepreview.com/?url=https://dotcms.com

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by manuelrojas**
### Summary
🐦🛠️🌐

The pull request adds support for twitter meta tags in the SEO tool. It defines new types, enums, and variables to handle the twitter tag values and preview data. It modifies the DotSeoMetaTagsService and the DotResultsSeoToolComponent to query, store, validate, and display the twitter meta tags in the page document. It also updates the templates, styles, and tests for the SEO tool components. It adds new messages to the Language.properties file and the DotMessageService mock for the twitter tag validation.

> _`Twitter` tags added_
> _SEO tool gets improved_
> _Autumn of refactor_

### Walkthrough
*  Add support for twitter meta tags in the SEO tool ([link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5L9-R13), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5L27-R33), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5R73-R76), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5L72-R88), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5L80-R99), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5R113-R132), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-e2c6edc96a3d4a7581030dec6768557e11cfc76d6863da2ff346f2cb3efbc715R56-R61), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-e2c6edc96a3d4a7581030dec6768557e11cfc76d6863da2ff346f2cb3efbc715R69-R72), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-e2c6edc96a3d4a7581030dec6768557e11cfc76d6863da2ff346f2cb3efbc715R142-R165), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-e2c6edc96a3d4a7581030dec6768557e11cfc76d6863da2ff346f2cb3efbc715R412-R594), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-89e37243c442c5e79f19f64f75095ce16f02aff9b8befc8f0af091039aabbff1R30-R45), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-07d21ddb95ed132c23e64c908dea0c11193cf632cebf2359b2cc4ca67709429bR126-R138), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-6084f02c375b9edee18f429040e76a983a8469637f28dd29a9c8fa330774fb88L73-R84), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-d290baa51a6e580d34ffe887fb5755d18dde6ea3657af381622595ec0f751104L173-R232), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-21db9723fa4c52389731b7575f1e19bd364aa992b7e549e22c08a33456994e94L5443-R5473))
*  Simplify the filtering and sorting of the seo results by the seoMedia input using the SeoMediaKeys object and the map operator ([link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5L72-R88), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5L80-R99), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-e2c6edc96a3d4a7581030dec6768557e11cfc76d6863da2ff346f2cb3efbc715L143-R184), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-6084f02c375b9edee18f429040e76a983a8469637f28dd29a9c8fa330774fb88L83-R103))
*  Store the preview data for all platforms in the allPreview property and use the MetaTagsPreview type ([link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-0b1cbca20fc4f4b1ccb288bb32c6dd4e26d1f4dc6d7647e530ca3c52dea1f1f5R113-R132),  [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-6084f02c375b9edee18f429040e76a983a8469637f28dd29a9c8fa330774fb88L66-R72))
*  Initialize and update the mainPreview property with the preview data for the selected platform ( [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-6084f02c375b9edee18f429040e76a983a8469637f28dd29a9c8fa330774fb88L83-R103))
*  Change the font-size of the dot-tab-buttons component to $spacing-3 for readability and consistency ([link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-b13e16f46401342cd8ea9f58eb1f995ed72327e7dc72d3287dbf68b36ee61925R11), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-b13e16f46401342cd8ea9f58eb1f995ed72327e7dc72d3287dbf68b36ee61925R21-R24))
*  Add data-testId attributes to the preview card elements for testing purposes ([link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-89e37243c442c5e79f19f64f75095ce16f02aff9b8befc8f0af091039aabbff1L21-R13), [link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-89e37243c442c5e79f19f64f75095ce16f02aff9b8befc8f0af091039aabbff1L27-R21))
*  Add mock messages for the DotMessageService to test the DotSeoMetaTagsService ([link](https://github.com/dotCMS/core/pull/26088/files?diff=unified&w=0#diff-f58cdbe27f1c14d0eb0ce89dec3179e958e29561996358866889c3a1ce122185L66-R102))

### Checklist
- [x] Tests
- [x] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
image

### CODE REVIEW

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.ts [hunk 1]
```diff
@@ -211,29 +234,48 @@ export class DotSeoMetaTagsService {
...
+        if (ogDescription?.length > SEO_LIMITS.MAX_OG_DESCRIPTION_LENGTH) {
+            result.push(
+                this.getWarningItem(this.dotMessageService.get('seo.rules.og-description.greater'))
+            );
+        }
+
+        if (
+            ogDescription &&
+            ogDescription?.length > SEO_LIMITS.MIN_OG_DESCRIPTION_LENGTH &&
+            ogDescription?.length > SEO_LIMITS.MAX_OG_DESCRIPTION_LENGTH
```
Comment (rjvelazco) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.ts:274]: Shouldn't this be
```suggestion
            ogDescription?.length <= SEO_LIMITS.MAX_OG_DESCRIPTION_LENGTH
```

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.ts [hunk 2]
```diff
@@ -53,13 +53,27 @@ export class DotSeoMetaTagsService {
...
         metaTagsObject['titleElements'] = title;
         metaTagsObject['favicon'] = (favicon[0] as HTMLLinkElement)?.href;
         metaTagsObject['title'] = title[0]?.innerText;
         metaTagsObject['titleOgElements'] = titleOgElements;
         metaTagsObject['imageOgElements'] = imagesOgElements;
+        metaTagsObject['twitterCardElements'] = twitterCardElements;
+        metaTagsObject['twitterTitleElements'] = twitterTitleElements;
+        metaTagsObject['twitterDescriptionElements'] = twitterDescriptionElements;
+        metaTagsObject['twitterImageElements'] = twitterImageElements;
+        metaTagsObject['descriptionOgElements'] = descriptionOgElements;
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.ts:76]: Does this `metaTagsObject` have a interface?

**Review by rjvelazco — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.html [hunk 3]
```diff
@@ -35,6 +27,22 @@ <h5>{{ mainPreview[0].title }}</h5>
                         </div>
                     </div>
                 </div>
+                <div *ngSwitchCase="seoMediaTypes.TWITTER">
+                    <img [src]="mainPreview.image" width="100%" alt="preview image" />
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.html:31]: img tags can't have units.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#width

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.ts [hunk 4]
```diff
@@ -80,12 +91,16 @@ export class DotResultsSeoToolComponent implements OnInit, OnChanges {
...
+      
+        const [preview] = this.allPreview;
+        this.mainPreview = preview;
     }
 
     ngOnChanges() {
-        this.currentResults = this.dotSeoMetaTagsService.getFilteredMetaTagsByMedia(
-            this.seoOGTagsResults,
-            this.seoMedia
+        this.currentResults = this.seoOGTagsResults.pipe(
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.ts:100]:
```suggestion
        this.currentResults$ = this.seoOGTagsResults.pipe(
```

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.ts (hunk 2 above)
Comment (manuelrojas) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.ts:76]: Yes it does: export interface SeoMetaTags

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.spec.ts [hunk 5]
```diff
@@ -102,18 +100,38 @@ describe('DotResultsSeoToolComponent', () => {
...
+        spectator.setInput({
+            seoMedia: SEO_MEDIA_TYPES.FACEBOOK
+        });
+        spectator.detectChanges();
+        spectator.component.currentResults$.subscribe((items) => {
             expect(items.length).toEqual(3);
             expect(items[0].key).toEqual(seoOGTagsResultMock[1].key);
             expect(items[1].key).toEqual(seoOGTagsResultMock[3].key);
             expect(items[2].key).toEqual(seoOGTagsResultMock[4].key);
         });
```
Comment (zJaaal) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.spec.ts:121]: Can you break this test? I think we had several problems in templates and experiments with this type of expect. If you cannot break it, you need to use the `done()` callback to make it work

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.spec.ts [hunk 6]
```diff
@@ -102,18 +100,38 @@ describe('DotResultsSeoToolComponent', () => {
...
+            seoMedia: SEO_MEDIA_TYPES.TWITTER
+        });
+        spectator.detectChanges();
+        spectator.component.currentResults$.subscribe((items) => {
+            expect(items.length).toEqual(4);
+            expect(items[0].key).toEqual(seoOGTagsResultOgMockTwitter[0].key);
+            expect(items[1].key).toEqual(seoOGTagsResultOgMockTwitter[1].key);
+            expect(items[2].key).toEqual(seoOGTagsResultOgMockTwitter[2].key);
+            expect(items[3].key).toEqual(seoOGTagsResultOgMockTwitter[3].key);
+        });
```
Comment (zJaaal) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.spec.ts:135]: Same here and all the expects that are async

**Review by zJaaal — COMMENTED** [review]
See my comments, let's be sure that the tests are working correctly

**Review by hmoreras — APPROVED** [review]

**Review by fmontes — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.spec.ts [hunk 7]
```diff
@@ -113,33 +165,79 @@ describe('DotSetMetaTagsService', () => {
...
+        ogMetaDescriptionSecond.setAttribute('content', 'Costa Rica Special Offer');
+
+        testDoc.head.appendChild(ogMetaDescription);
+        testDoc.head.appendChild(ogMetaDescriptionSecond);
+
+        service.getMetaTagsResults(testDoc).subscribe((value) => {
+            expect(value[0].items[0].message).toEqual('more than 1 og:description meta tag found!');
+            expect(value[0].items[1].message).toEqual(
+                'og:description meta tag found, but has fewer than 55 characters of content.'
+            );
```
Comment (oidacra) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.spec.ts:220]: Missing done()

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.spec.ts [hunk 8]
```diff
@@ -113,33 +165,79 @@ describe('DotSetMetaTagsService', () => {
...
+        ogMetaTitle.setAttribute('content', 'Costa Rica Special Offer');
+
+        const ogMetaTitleSecond = document.createElement('meta');
+        ogMetaTitleSecond.setAttribute('property', 'og:title');
+        ogMetaTitleSecond.setAttribute('content', 'Costa Rica Special Offer');
+
+        testDoc.head.appendChild(ogMetaTitle);
+        testDoc.head.appendChild(ogMetaTitleSecond);
+
+        service.getMetaTagsResults(testDoc).subscribe((value) => {
```
Comment (oidacra) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/content/services/html/dot-seo-meta-tags.service.spec.ts:236]: same here done()

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.spec.ts [hunk 9]
```diff
@@ -102,18 +100,38 @@ describe('DotResultsSeoToolComponent', () => {
...
-    it('should filter seo results by seoMedia on changes', () => {
-        spectator.fixture.componentInstance.seoMedia = 'facebook';
-        spectator.component.ngOnChanges();
-        spectator.component.currentResults.subscribe((items) => {
+    it('should filter seo results by Facebook, seoMedia on changes', () => {
+        spectator.setInput({
+            seoMedia: SEO_MEDIA_TYPES.FACEBOOK
+        });
+        spectator.detectChanges();
+        spectator.component.currentResults$.subscribe((items) => {
```
Comment (oidacra) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/seo/components/dot-results-seo-tool/dot-results-seo-tool.component.spec.ts:116]: remember to use done() when you subscribe in a test. If not can have false positives.

**Review by oidacra — APPROVED** [review]

**Pull request merged**